src/shared/ui/pickerPrompter.ts

```diff
@@ -0,0 +1,255 @@
+    ignoreFocusOut: true,
+}
+
+type QuickPickData<T> = PromptResult<T> | (() => Promise<PromptResult<T>>)
+type LabelQuickPickItem<T> = vscode.QuickPickItem & { label: T }
+
+/**
+ * Attaches additional information as `data` to a QuickPickItem. Alternatively, `data` can be a function that
+ * returns a Promise, evaluated after the user selects the item.
+ */
+export type DataQuickPickItem<T> = vscode.QuickPickItem & { data: QuickPickData<T> }
+export type DataQuickPick<T> = Omit<vscode.QuickPick<DataQuickPickItem<T>>, 'buttons'> & { buttons: PrompterButtons<T> }
+
+const CUSTOM_USER_INPUT = Symbol()
+
+function isDataQuickPickItem(obj: any): obj is DataQuickPickItem<any> {
+    return typeof obj === 'object' && typeof (obj as vscode.QuickPickItem).label === 'string' && 'data' in obj
+}
+
+/**
+ * Creates a new QuickPick using special DataQuickPickItem interfaces. Information that should be returned when
+ * the user selects an item should be placed in the `data` property of each item. If only the `label` is desired,
+ * use `createLabelQuickPick` instead.
+ *
+ * @param items An array or a Promise for items.
+ * @param options Customizes the QuickPick and QuickPickPrompter.
+ * @returns A QuickPickPrompter. This can be used directly with the `prompt` method or can be fed into a Wizard.
+ */
+export function createQuickPick<T>(
+    items: DataQuickPickItem<T>[] | Promise<DataQuickPickItem<T>[] | undefined>,
+    options?: ExtendedQuickPickOptions<T>
+): QuickPickPrompter<T> {
+    const picker = vscode.window.createQuickPick<DataQuickPickItem<T>>() as DataQuickPick<T>
+    options = { ...DEFAULT_QUICKPICK_OPTIONS, ...options }
+    applyPrimitives(picker, { ...DEFAULT_QUICKPICK_OPTIONS, ...options })
+
+    const prompter =
+        options.customInputSettings !== undefined
+            ? new CustomQuickPickPrompter<T>(
+                  picker,
+                  options.customInputSettings.label,
+                  options.customInputSettings.transform
+              )
+            : new QuickPickPrompter<T>(picker)
+
+    if (items instanceof Promise) {
+        makePickerAysnc(picker, items)
+    } else {
+        picker.items = items
+    }
+
+    return prompter
+}
+
+/** Creates a QuickPick from normal QuickPickItems, using the `label` as the return value. */
+export function createLabelQuickPick<T extends string>(
+    items: LabelQuickPickItem<T>[] | Promise<LabelQuickPickItem<T>[] | undefined>,
+    options?: ExtendedQuickPickOptions<T>
+): QuickPickPrompter<T> {
+    if (items instanceof Promise) {
+        return createQuickPick(
+            items.then(items => (items !== undefined ? items.map(item => ({ ...item, data: item.label })) : undefined)),
+            options
+        )
+    }
+    return createQuickPick(
+        items.map(item => ({ ...item, data: item.label })),
+        options
+    )
+}
+
+function makePickerAysnc<T>(picker: DataQuickPick<T>, items: Promise<DataQuickPickItem<T>[] | undefined>): void {
```

Comment:
what's async about this? a 1-line docstring would help.

Reply:
Ah yeah bad naming on my part. The function is making an already existing picker load items asynchronously, not creating a new picker asynchronously. I'm moving this logic into the class since it doesn't really need to be separate.